docs/source/usrman/mpi4py.util.pkl5.rst

```diff
@@ -0,0 +1,226 @@
+mpi4py.util.pkl5
+-----------------
+
+.. module:: mpi4py.util.pkl5
+   :synopsis: Pickle-based communication using protocol 5.
+
+.. versionadded:: 3.1.0
+
+:mod:`pickle` protocol 5 (see :pep:`574`) introduced support for out-of-band
+buffers, allowing for more efficient handling of certain object types with
+large memory footprints.
+
+MPI for Python uses the traditional in-band handling of buffers. This approach
+is appropriate for communicating non-buffer Python objects, or buffer-like
+objects with small memory footprints. For point-to-point communication, in-band
+buffer handling allows for the communication of a pickled stream with a single
+MPI message, at the expense of additional CPU and memory overhead in the
+pickling and unpickling steps.
+
+The :mod:`mpi4py.util.pkl5` module provides communicator wrapper classes
+reimplementing pickle-based point-to-point communication methods using pickle
+protocol 5. Handling out-of-band buffers necessarily involve multiple MPI
+messages, thus increasing latency and hurting performance in case of small size
+data. However, in case of large size data, the zero-copy savings of out-of-band
+buffer handling more than offset the extra latency costs.  Additionally, these
+wrapper methods overcome the infamous 2 GiB MPI-3 message count limit.
+
+.. note::
+
+   Support for pickle protocol 5 is available in the :mod:`pickle` module
+   within the Python standard library since Python 3.8. Previous Python 3
+   releases can use the :mod:`pickle5` backport, which is available on `PyPI
+   <pickle5-pypi_>`_ and can be installed with::
+
+       python -m pip install pickle5
+
+   .. _pickle5-pypi: https://pypi.org/project/pickle5/
```

Comment:
I said *... previous Python **3** releases ...* and not *... previous Python releases ...*. 

So an HPC group (HPC!!! what a surprise!!!) don't want to pay the price of moving on while keep spending our time supporting them? We give them working code. I'm not willing to spend my time documenting/clarifying Python 2 support, not after Python 2 is officially dead.

Reply:
Well, if I understand you correctly, then unfortunately we still have a one-time cost to pay. AFAIK we never declared mpi4py is Python 3 only -- at least it's not in the 3.0.2 announcement in the mailing list (and I don't find one for 3.0.3), and last time I checked we were still inquiring feedbacks about Python 2.7: https://groups.google.com/g/mpi4py/c/-FufUsS3jOU/m/BXm-nPxDEgAJ. I think for the purpose here it wouldn't hurt to add one sentence along this line "this feature is not supported on Python 2".

Reply:
Yes, you asked for feedback, a not a single soul cared to answer. What should we take of that? Python 2 is effectively dead :wink:?  OK, I'll add a note saying that while the code runs fine on Python 2, the lack of pickle 5 and other limitations of such an old runtime gives no actual benefit (no out-of-band communication, no large-array support).

Reply:
> Yes, you asked for feedback, a not a single soul cared to answer. What should we take of that? Python 2 is effectively dead 😉?

Yeah it is both sad and good that no one replies...Sad because of the little "community" engagement, and good because we can do whatever we'd like to 🎉

Reply:
I'm going to extend the `pickle5` note with the following sentences. Are you OK with it?

```
   The :mod:`pickle5` backport is not available for Python 2. Using the
   :mod:`mpi4py.util.pkl5` module on a Python 2 runtime provides no benefits
   and may hurt communication performance.
```

Reply:
LGTM!